### PR TITLE
Add question validations to backend

### PR DESCRIPTION
**Pull request by drewpotter**
This is the initial work to add validations to the backend, with frontend validations in place too. I have disabled one of the tests because of an issue with making the test pass which has not been resolved and this will be fixed later on.

### CODE REVIEW

File: app/controllers/responses_controller.rb [hunk 1]
```diff
@@ -24,19 +24,25 @@ def new
...
+    answer_params = validate_responses(response_params)
+
+    if answer_params.present?
+      @response            = Response.new(answer_params)
+      @response.survey     = @survey
+      @response.ip_address = request.remote_ip
+      @response.language   = params[:locale]
+
+      respond_to do |format|
+        if @response.save && answer_params.present?
```
Comment (spencerldixon) [inline app/controllers/responses_controller.rb]: The `answer_params.present?` is redundant here as it's already being checked in the enclosing if statement

File: app/controllers/responses_controller.rb [hunk 2]
```diff
@@ -24,19 +24,25 @@ def new
...
+          format.html { redirect_to root_path, notice: 'Response was successfully created.' }
+          format.json { render :show, status: :created, location: @response }
+        else
+          format.html { render :new }
+          format.json { render json: @response.errors, status: :unprocessable_entity }
+        end
       end
+    else
```
Comment (spencerldixon) [inline app/controllers/responses_controller.rb]: I would probably flip this around, `redirect_to path unless answer_params.present?`, you should be able to get rid of the large enclosing conditional

File: app/controllers/responses_controller.rb [hunk 3]
```diff
@@ -68,6 +74,21 @@ def response_params
       params.require(:response).permit(answers_attributes: [{raw: []}, :raw, :answerable_type, :answerable_id])
     end
 
+    def validate_responses(params)
+      answers = params["answers_attributes"].values
+
+      required_questions = Question.pluck(:id)
+      required_demographic_questions = DemographicQuestion.where(validation: {required: true}.to_json).pluck(:id)
```
Comment (spencerldixon) [inline app/controllers/responses_controller.rb]: Would make sense to move these to the actual objects so you could simply do `Question.required_questions` or something similar for both

File: app/controllers/responses_controller.rb (hunk 1 above)
Comment (drewpotter) [reply app/controllers/responses_controller.rb]: Thanks well spotted!

File: app/controllers/responses_controller.rb (hunk 2 above)
Comment (drewpotter) [reply app/controllers/responses_controller.rb]: Many thanks!

File: app/controllers/responses_controller.rb (hunk 3 above)
Comment (drewpotter) [reply app/controllers/responses_controller.rb]: Done now!

File: app/controllers/responses_controller.rb [hunk 4]
```diff
@@ -68,6 +71,21 @@ def response_params
       params.require(:response).permit(answers_attributes: [{raw: []}, :raw, :answerable_type, :answerable_id])
     end
 
+    def validate_responses(params)
```
Comment (spencerldixon) [inline app/controllers/responses_controller.rb]: Can you explain this method to me?

Comment (drewpotter) [reply app/controllers/responses_controller.rb]: Sure, this method will either return the params, if the required questions have all been answered or nil if there are any issues - so missing answered questions. In reality, due to the frontend validations there is no way someone could possibly submit the wrong questions through the website itself. So this is more of a protection against for example bugs elsewhere.

Comment (spencerldixon) [reply app/controllers/responses_controller.rb]: Hmm ok, I think we could make this a little simpler, for example, we could raise the error and add it to the Response object, which would be a much more rails way of rejecting invalid params. I know we might be pressed for time so I can merge this if you like, and we can come back to it later on perhaps?

Comment (drewpotter) [reply app/controllers/responses_controller.rb]: Please don't merge this yet! I will try and rework this to make it better.

**Pull request merged**